trace-obfuscation/src/replacer.rs

```diff
@@ -0,0 +1,188 @@
+}
+
+/// replace_trace_tags replaces the tag values of all spans within a trace with a given set of rules.
+pub fn replace_trace_tags(trace: &mut [pb::Span], rules: &[ReplaceRule]) {
+    for rule in rules {
+        for span in &mut *trace {
+            match rule.name {
+                "*" => {
+                    for (_, val) in span.meta.iter_mut() {
+                        *val = rule.re.replace_all(val, rule.repl).to_string();
+                    }
+                }
+                "resource.name" => {
+                    span.resource = rule.re.replace_all(&span.resource, rule.repl).to_string();
+                }
+                _ => {
+                    if let Some(val) = span.meta.get_mut(rule.name) {
+                        let replaced_tag = rule.re.replace_all(val, rule.repl).to_string();
+                        *val = replaced_tag;
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// parse_rules_from_string takes an array of rules, represented as an array of length 3 arrays
+/// holding the tag name, regex pattern, and replacement string as strings.
+/// * returns a vec of ReplaceRules
+pub fn parse_rules_from_string<'a>(
+    rules: &'a [[&'a str; 3]],
+) -> anyhow::Result<Vec<ReplaceRule<'a>>> {
+    let mut vec: Vec<ReplaceRule> = Vec::with_capacity(rules.len());
+
+    for [name, pattern, repl] in rules {
+        let compiled_regex = match Regex::new(pattern) {
+            Ok(res) => res,
+            Err(err) => {
+                anyhow::bail!(format!(
+                    "Obfuscator Error: Error while parsing rule: {}",
+                    err
+                ))
+            }
+        };
+        vec.push(ReplaceRule {
+            name,
+            re: compiled_regex,
+            repl,
+        });
+    }
+    Ok(vec)
+}
+
+#[cfg(test)]
+mod tests {
+
+    use crate::pb;
+    use crate::replacer;
+    use duplicate::duplicate_item;
+    use std::collections::HashMap;
+
+    fn new_test_span_with_tags(tags: HashMap<&str, &str>) -> pb::Span {
+        let mut span = pb::Span {
+            duration: 10000000,
+            error: 0,
+            resource: "GET /some/raclette".to_string(),
+            service: "django".to_string(),
+            name: "django.controller".to_string(),
+            span_id: 123,
+            start: 1448466874000000000,
+            trace_id: 424242,
+            meta: HashMap::new(),
+            metrics: HashMap::from([("cheese_weight".to_string(), 100000.0)]),
+            parent_id: 1111,
+            r#type: "http".to_string(),
+            meta_struct: HashMap::new(),
+        };
+        for (key, val) in tags {
+            match key {
+                "resource.name" => {
+                    span.resource = val.to_string();
+                }
+                _ => {
+                    span.meta.insert(key.to_string(), val.to_string());
+                }
+            }
+        }
+        span
+    }
+
+    #[duplicate_item(
```

Comment:
I've mixed feelings personally for using heavy meta programming here. Though mostly around having to deal with IDE integration and automated formatting

Though it does save on some boilerplate - and looks somewhat nice :shrug: 

Just my 3c